dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateVectorStoreRecordCollection.cs

```diff
@@ -0,0 +1,383 @@
+        typeof(int?),
+        typeof(long),
+        typeof(long?),
+        typeof(short),
+        typeof(short?),
+        typeof(byte),
+        typeof(byte?),
+        typeof(float),
+        typeof(float?),
+        typeof(double),
+        typeof(double?),
+        typeof(decimal),
+        typeof(decimal?),
+        typeof(DateTime),
+        typeof(DateTime?),
+        typeof(DateTimeOffset),
+        typeof(DateTimeOffset?),
+        typeof(Guid),
+        typeof(Guid?),
+        typeof(bool),
+        typeof(bool?)
+    ];
+
+    /// <summary>Default JSON serializer options.</summary>
+    private static readonly JsonSerializerOptions s_jsonSerializerOptions = new()
+    {
+        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
+        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
+        Converters =
+        {
+            new WeaviateDateTimeOffsetConverter(),
+            new WeaviateNullableDateTimeOffsetConverter()
+        }
+    };
+
+    /// <summary><see cref="HttpClient"/> that is used to interact with Weaviate API.</summary>
+    private readonly HttpClient _httpClient;
+
+    /// <summary>Optional configuration options for this class.</summary>
+    private readonly WeaviateVectorStoreRecordCollectionOptions<TRecord> _options;
+
+    /// <summary>A definition of the current storage model.</summary>
+    private readonly VectorStoreRecordDefinition _vectorStoreRecordDefinition;
+
+    /// <summary>A dictionary that maps from a property name to the storage name that should be used when serializing it to JSON for data and vector properties.</summary>
+    private readonly Dictionary<string, string> _storagePropertyNames;
+
+    /// <summary>The key property of the current storage model.</summary>
+    private readonly VectorStoreRecordKeyProperty _keyProperty;
+
+    /// <summary>The data properties of the current storage model.</summary>
+    private readonly List<VectorStoreRecordDataProperty> _dataProperties;
+
+    /// <summary>The vector properties of the current storage model.</summary>
+    private readonly List<VectorStoreRecordVectorProperty> _vectorProperties;
+
+    /// <summary>The mapper to use when mapping between the consumer data model and the Weaviate record.</summary>
+    private readonly IVectorStoreRecordMapper<TRecord, JsonNode> _mapper;
+
+    /// <inheritdoc />
+    public string CollectionName { get; }
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="WeaviateVectorStoreRecordCollection{TRecord}"/> class.
+    /// </summary>
+    /// <param name="httpClient"><see cref="HttpClient"/> that is used to interact with Weaviate API.</param>
+    /// <param name="collectionName">The name of the collection that this <see cref="WeaviateVectorStoreRecordCollection{TRecord}"/> will access.</param>
+    /// <param name="options">Optional configuration options for this class.</param>
+    public WeaviateVectorStoreRecordCollection(
+        HttpClient httpClient,
```

Comment:
Curious about authentication here also?

If the `HttpClient` needs to be configured with authentication options then we should make this very clear in the API documentation. An option to pass in the API key might be nice also as we do this elsewhere in our API.

Reply:
Added a possibility to specify API key in `options` and added XML documentation for `HttpClient` that it can also be included via `DefaultRequestHeaders` property. Also added tests to verify it is used correctly.